### PR TITLE
feat!: create a server version of peacock

### PR DESCRIPTION
**Pull request by Skisocks**
## Changes
This PR introduces the peacock service. 
Currently it supports all the same features of the CLI.

### CODE REVIEW

File: .lighthouse/jenkins-x/peacock/verify.yaml [hunk 1]
```diff
@@ -20,11 +20,12 @@ spec:
               script: |
                 #!/usr/bin/env sh
                 jx gitops variables --commit=false
-                echo "export GITHUB_TOKEN=\"$(kubectl get secret peacock -n jx-staging -o jsonpath="{.data['gh-bot-token']}" | base64 -d)\"" >> .jx/variables.sh
+
```
Comment (tomhobson) [inline .lighthouse/jenkins-x/peacock/verify.yaml:23]: whitespace

**Review by tomhobson — CHANGES_REQUESTED** [review]

File: .lighthouse/jenkins-x/peacock/release.yaml [hunk 2]
```diff
@@ -20,11 +20,12 @@ spec:
               script: |
                 #!/usr/bin/env sh
                 jx gitops variables --commit=false
-                echo "export GITHUB_TOKEN=\"$(kubectl get secret peacock -n jx-staging -o jsonpath="{.data['gh-bot-token']}" | base64 -d)\"" >> .jx/variables.sh
+
```
Comment (tomhobson) [inline .lighthouse/jenkins-x/peacock/release.yaml:23]: Whitespace

**Review by tomhobson — CHANGES_REQUESTED** [review]

File: charts/peacock/Chart.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+apiVersion: v1
+description: A Helm chart for Kubernetes
+icon: https://raw.githubusercontent.com/cdfoundation/artwork/master/jenkinsx/icon/color/jenkinsx-icon-color.png
+name: peacock
+version: 0.1.0-SNAPSHOT
```
Comment (tomhobson) [inline charts/peacock/Chart.yaml:5]: Should we get the image rocking and rolling?

File: charts/peacock/values.yaml [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    # the domain for hosts
+    domain: ""
+    externalDNS: false
+    namespaceSubDomain: -jx.
+    serviceType: ""
+    tls:
+      email: ""
+      enabled: false
+      production: false
+      secretName: ""
```
Comment (tomhobson) [inline charts/peacock/values.yaml:118]: does this make it jx native?

File: pkg/cmd/run/run.go [hunk 5]
```diff
@@ -46,10 +40,10 @@ type Options struct {
 	CommentValidation bool
 	Subject           string
 
-	GitServerClient domain.GitServer
+	GitServerClient domain.SCM
 	Git             domain.Git
-	Handlers        map[string]domain.MessageHandler
-	Config          *feathers.Feathers
+	MSGHandler      domain.MessageHandler
```
Comment (tomhobson) [inline pkg/cmd/run/run.go:45]: MessageHandler?

File: pkg/git/github/client.go [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+	prsWithCommit, _, err := c.github.PullRequests.ListPullRequestsWithCommit(ctx, c.owner, c.repo, sha, nil)
+	if err != nil {
+		return nil, err
+	}
+	if len(prsWithCommit) < 1 {
+		return nil, errors.Errorf("no pull request found containing commit %s", sha)
+	}
+	log.Infof("Found %d pull request(s) containing that commit", len(prsWithCommit))
+
+	// If there is only one PR then that must be is
```
Comment (tomhobson) [inline pkg/git/github/client.go:70]: `Then that must be is`

File: pkg/health/health.go [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+// ServeHealth godoc
+// @Summary Gets the health of the service
+// @Description get the health of the dependencies of the service
+// @Tags Health
+// @Produce json
+// @Success 200
+// @Router /health [get]
+func ServeHealth(c *gin.Context) {
+	// Todo: Add health checks for other dependencies
```
Comment (tomhobson) [inline pkg/health/health.go:15]: :) Fine because this a POC, but we should maybe add some health stuff

File: pkg/message/message.go [hunk 8]
```diff
@@ -53,9 +59,54 @@ func ParseTeamNames(teamNameReg *regexp.Regexp, markdown string) [][]string {
...
+Message {{ inc $idx }} will be sent to: {{ commaSep $val.TeamNames }}
+<details>
+<summary>Message Breakdown</summary>
+
+{{ $val.Content }}
+
+</details>
+
+{{ end -}}`
+
```
Comment (tomhobson) [inline pkg/message/message.go:92]: Feel like this should be in a config or a constant at least

File: pkg/msgclient/handler.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+func NewMessageHandler(cfg *config.MessageHandlers) *Handler {
+	clients := make(map[string]domain.MessageClient)
+	if cfg.Slack.Token != "" {
+		log.Info("Slack message handler initialised")
+		clients[models.Slack] = slack.NewClient(cfg.Slack.Token)
+	}
+	if cfg.Webhook.URL != "" && cfg.Webhook.Secret != "" {
+		log.Info("Webhook message handler initialised")
+		clients[models.Webhook] = webhook.NewClient(cfg.Webhook.URL, cfg.Webhook.Token, cfg.Webhook.Secret)
```
Comment (tomhobson) [inline pkg/msgclient/handler.go:28]: What happens if they are empty?

File: pkg/server/data.go [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+
+import "github.com/spring-financial-group/peacock/pkg/config"
+
+type DataSources struct {
+}
+
+// initDataSources creates the data sources for the server, initialising Redis, Postgres, etc.
+func initDataSources(cfg *config.Config) (*DataSources, error) {
+	return &DataSources{}, nil
+}
```
Comment (tomhobson) [inline pkg/server/data.go:11]: Is this just a proxy method? I'm not really sure what this is doing?

**Review by tomhobson — CHANGES_REQUESTED** [review]

**Review by tomhobson — APPROVED** [review]

**Comment by Skisocks** [comment]
wut
/test peacock-verify

**Comment by Skisocks** [comment]
Did I push the git token?

**Comment by Skisocks** [comment]
Secrets must have been having a nap

**Pull request merged**